refactoring TestQuantizeScript

edmundw314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39677 refactoring TestQuantizeScript**

Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

edmundw314 committed:
refactoring TestQuantizeScript

[ghstack-poisoned]

edmundw314 committed:
Update on "refactoring TestQuantizeScript"


Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

[ghstack-poisoned]

edmundw314 committed:
Update on "refactoring TestQuantizeScript"


Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

[ghstack-poisoned]

edmundw314 committed:
Update on "refactoring TestQuantizeScript"


Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -7,29 +7,55 @@
+    default_per_channel_weight_observer,
+    default_qconfig,
+    get_default_qconfig,
+
```
remove empty line?

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -7,29 +7,55 @@
+    default_per_channel_qconfig,
+    test_only_eval_fn,
+)
+#annotated models
```
lint

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2574,3 +2600,231 @@ def forward(self, x):
+        r"""Compare the result of quantizing single linear layer in
+        eager mode and graph mode
+        """
+        for qengine in supported_qengines:
+            with override_quantized_engine(qengine):
```
can you use this decorator: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_quantized.py#L119

you can search in the code base to see how this is used.

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2574,3 +2600,231 @@ def forward(self, x):
+                        inplace=False)
+                    self.assertEqual(model_quantized(self.img_data[0][0]), result_eager)
+
+    @unittest.skip("This doesn't work right now, re-enable after fold_convbn is fixed")
```
try enabling this?

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2574,3 +2600,231 @@ def forward(self, x):
+        self.assertEqual(result_eager, result_script)
+
+    def test_nested(self):
+        for qengine in supported_qengines:
+            with override_quantized_engine(qengine):
```
same here, we can use decorator and remove this

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2574,3 +2600,231 @@ def forward(self, x):
+    def test_skip_quant(self):
+        """ Test None qconfig
+        """
+        for qengine in supported_qengines:
+            with override_quantized_engine(qengine):
```
same here

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2574,3 +2600,231 @@ def forward(self, x):
+        r"""Compare the result of dynamic quantization of single linear layer in
+        eager mode and graph mode.
+        """
+        if 'qnnpack' in supported_qengines:
+            with override_quantized_engine('qnnpack'):
```
you can use decorator and then check if the current engine is qnnpack, you can add a similar function: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_quantized.py#L127

jerryzh168 commented on test/test_quantization.py:
```diff
@@ -54,6 +53,8 @@
 from quantization.test_quantize_script import TestQuantizeScriptJitPasses  # noqa: F401
 from quantization.test_quantize_script import TestQuantizeScriptPTSQOps  # noqa: F401
 from quantization.test_quantize_script import TestQuantizeDynamicScript  # noqa: F401
+# TODO: move to test_quantize_script
```
remove comment

edmundw314 committed:
Update on "refactoring TestQuantizeScript"


Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

[ghstack-poisoned]

edmundw314 committed:
Update on "refactoring TestQuantizeScript"


Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -7,31 +7,60 @@
 # Testing utils
-from torch.testing._internal.common_quantization import test_only_eval_fn as _test_only_eval_fn
-from torch.testing._internal.common_quantized import override_qengines
+from torch.testing._internal.common_quantized import (
```
lint

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2625,3 +2654,224 @@ def forward(self, x):
         for x, obs in model._modules._c.items():
             graph_params.append((obs.getattr('3_scale_0'), obs.getattr('3_zero_point_0')))
         self.assertEqual(ref_qparams, graph_params)
+class TestQuantizeScript(QuantizationTestCase):
```
please include new line before the class, wonder why there is no lint for this one

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2625,3 +2654,224 @@ def forward(self, x):
+        linear_model.fc1.weight = torch.nn.Parameter(annotated_linear_model.fc1.module.weight.detach())
+        linear_model.fc1.bias = torch.nn.Parameter(annotated_linear_model.fc1.module.bias.detach())
+        model_eager = quantize(annotated_linear_model, test_only_eval_fn,
+                                self.calib_data)
```
lint

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2625,3 +2654,224 @@ def forward(self, x):
+                inplace=False)
+            self.assertEqual(model_quantized(self.img_data[0][0]), result_eager)
+
+    # @unittest.skip("This doesn't work right now, re-enable after fold_convbn is fixed")
```
this can be removed now?

jerryzh168 reviewed (commented):
Please address lint comments

jerryzh168 reviewed (changes requested):
please address the lint and some inline comments

edmundw314 committed:
Update on "refactoring TestQuantizeScript"


Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

[ghstack-poisoned]

jerryzh168 commented:
Note there are still some lint errors in the PR, please take a look at https://github.com/pytorch/pytorch/pull/39677/files

edmundw314 committed:
Update on "refactoring TestQuantizeScript"


Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
Looks good, thanks!
Please also request review after you updated the PR

jerryzh168 commented:
OSS CI is failing, please make sure all CIs passes or the ones that failed is irrelevant to the PR before landing

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -7,31 +7,58 @@
+from torch.quantization import (
+    QConfig,
+    default_dynamic_qconfig,
+    default_observer,
```
probably because you missed per_channel_dynamic_qconfig here I think

edmundw314 committed:
Update on "refactoring TestQuantizeScript"


Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

[ghstack-poisoned]

edmundw314 committed:
Update on "refactoring TestQuantizeScript"


Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

[ghstack-poisoned]

edmundw314 committed:
Update on "refactoring TestQuantizeScript"


Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

[ghstack-poisoned]

edmundw314 committed:
Update on "refactoring TestQuantizeScript"


Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2584,41 +2643,238 @@ def forward(self, x):
-                else:
-                    graph_params.append((obs.getattr('3_scale_0'), obs.getattr('3_zero_point_0')))
-            self.assertEqual(ref_qparams, graph_params)
+        model = torch.jit.script(M()).eval()
+        qconfig = script_qconfig(default_dynamic_qconfig)
+        ref_qparams = []
+        wt_module = wrap_cpp_module(qconfig.weight)
+        for wt in [model.fc.weight, model.fc2.weight]:
+            wt_module(wt)
+            qparams = wt_module.calculate_qparams()
+            ref_qparams.append((qparams[0].item(), qparams[1].item()))
+        model = prepare_dynamic_script(model, qconfig_dict)
+        model = convert_dynamic_script(model, debug=True)
+        graph_params = []
+        for x, obs in model._modules._c.items():
+            graph_params.append((obs.getattr('3_scale_0'), obs.getattr('3_zero_point_0')))
+        self.assertEqual(ref_qparams, graph_params)
```
please revert these changes

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -1201,6 +1229,37 @@ class TestQuantizeScriptPTSQOps(QuantizationTestCase):
     """ Test graph mode post training static quantization works
     for individual ops end to end.
     """
+    def _test_op_impl(self, module, data, quantized_op, tracing=False, debug=False, check=True, eval_mode=True):
+        if debug:
+            print('Testing:', str(module))
+        qconfig_dict = {'': get_default_qconfig(torch.backends.quantized.engine)}
+        if eval_mode:
+            module = module.eval()
+        *inputs, target = data[0]
+        model = get_script_module(module, tracing, inputs).eval()
+        models = {}
+        outputs = {}
+        for d in [True, False]:
+            # TODO: _test_only_eval_fn --> default_eval_fn
+            models[d] = quantize_script(
+                model, qconfig_dict, test_only_eval_fn, [data], inplace=False, debug=d)
+            # make sure it runs
+            outputs[d] = models[d](*inputs)
+
+        if debug:
+            print('debug graph:', models[True].graph)
+            print('non debug graph:', models[False].graph)
+
```
also this

edmundw314 committed:
Update on "refactoring TestQuantizeScript"


Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

[ghstack-poisoned]

edmundw314 committed:
Update on "refactoring TestQuantizeScript"


Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

[ghstack-poisoned]

edmundw314 committed:
Update on "refactoring TestQuantizeScript"


Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

[ghstack-poisoned]

edmundw314 committed:
Update on "refactoring TestQuantizeScript"


Differential Revision: [D21940638](https://our.internmc.facebook.com/intern/diff/D21940638)

[ghstack-poisoned]
